### PR TITLE
[Debugger] Fixed bug where we loop randomly long(+1) and optimised stepping in(over) non-user code

### PR DESCRIPTION
**Pull request by DavidKarlas**

### CODE REVIEW

File: mono/mini/debugger-agent.c [hunk 1]
```diff
@@ -4911,6 +4913,22 @@ process_single_step_inner (DebuggerTlsData *tls, gboolean from_signal)
 		return;
 
 	/*
+	 * This could be in ss_update method, but mono_find_next_seq_point_for_native_offset is pretty expensive method,
+	 * hence we prefer this check here.
+	 */
+	if (ss_req->user_assemblies) {
+		gboolean found = FALSE;
+		for (int k = 0; ss_req->user_assemblies[k]; k++)
+			if (ss_req->user_assemblies[k] == jinfo_get_method(ji)->klass->image->assembly) {
```
Comment (Therzok) [inline mono/mini/debugger-agent.c]: jinfo_get_method seems like an invariant

**Comment by DavidKarlas** [comment]
Addressed comment from Marius

File: mono/mini/debugger-agent.c [hunk 2]
```diff
@@ -569,6 +569,8 @@ typedef struct {
 	GSList *bps;
 	/* The number of frames at the start of a step-over */
 	int nframes;
+	/* If set, don't stop in methods that are not part of this assemblies */
+	MonoAssembly** user_assemblies;
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: Typo. 'this assemblies'.

**Comment by vargaz** [comment]
Looks ok otherwise.

**Comment by DavidKarlas** [comment]
Changed comment from "this assemblies" to "user assemblies"

**Pull request merged**